ssl/statem/statem_clnt.c

```diff
@@ -821,7 +821,10 @@ int tls_construct_client_hello(SSL *s)
         goto err;
     }
 
-    if ((sess == NULL) || (sess->ssl_version != s->version) ||
+    if ((sess == NULL) ||
+        /* Only destroy session which version is bigger than maximum supported.
+         */
+        ssl_version_cmp(s, sess->ssl_version, s->version) > 0 ||
```

Comment:
@mattcaswell Servers should be doing version negotiation before session resumption and won't get into that situation. Any server that does this in the wrong order already cannot interoperate with other implementations. No client implementation (not even OpenSSL, due to this version locking) accepts a session/version mismatch, and most clients are not version-locked, so they will freely offer a TLS 1.1 session on a TLS 1.2 ClientHello.

Plus the particulars of session resumption are version-specific (TLS 1.3 is due to change it very significantly), so doing it in the wrong order isn't meaningful.